### DISCUSSION TITLE
"Value error, Expected UploadFile, received: <class 'str'>"

### DISCUSSION THREAD

**Discussion by markodjordjic** [fastapi/fastapi#11596] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import FastAPI, File, UploadFile

app = FastAPI()


@app.post("/upload_file/")
async def create_upload_file(file: UploadFile):
    
    return {"filename": file.filename}

def test_upload_file:

        with tempfile.SpooledTemporaryFile(mode='wb') as fp:
            fp.write(b'Hello World!')
            fp.seek(0)

        upload_file = UploadFile(file=fp, filename='test_file.txt')

        data = {
            "file": upload_file
        }
        
        response = app.post("/upload_file/", data=data)

        actual = response.json()
    
        self.assertEqual(actual, expected)
```

### Description

* Instead of reading the file, error in the title is thrown

### Operating System

Windows

### Operating System Details

Windows 11 Pro

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

3.11.0

**Comment by YuriiMotov** [comment]
Hi!
Why don't you use TestClient to test your endpoint?
```
import tempfile

from fastapi import FastAPI, UploadFile
from fastapi.testclient import TestClient

app = FastAPI()


@app.post("/upload_file/")
async def create_upload_file(file: UploadFile):
    return {"filename": file.filename}


client = TestClient(app)

def test_upload_file():
    file_name = 'test_file.txt'
    expected = {"filename": file_name}
    with tempfile.SpooledTemporaryFile(mode='wb+') as fp:
        fp.write(b'Hello World!')
        fp.seek(0)
        files = {'file': (file_name, fp)}
        response = client.post("/upload_file/", files=files)
        print(response.status_code)

    actual = response.json()

    assert actual == expected
    # self.assertEqual(actual, expected)
```

**Reply by markodjordjic** [reply]
@YuriiMotov if I understand correctly your proposal, you advocate for not sending the `UploadFile` class but a `dict` with a key `file` and a value that will hold a `tuple` with `file_name` and opened `file` itself?

**Reply by markodjordjic** [reply]
@YuriiMotov Regarding the absence of use of `TestClient` the example I posted is clumsy, since I extracted it from larger project. Your example is completely fine.

**Reply by markodjordjic** [reply]
@YuriiMotov I can confirm that the proposal you suggested is working. But that is a workaround without using the `UploadFile` class. I am sure this is a bug; since once the `UploadFile` class is there, the error is thrown.

**Reply by YuriiMotov** [reply, marked as answer]
My example shows a common way of testing FastAPI applications: using TestClient to imitate HTTP-request to your application.
```
response = client.post("/upload_file/", files=files)
```
Instead of
```
response = app.post("/upload_file/", data=data) # This is wrong!
```

In your code `app.post("/upload_file/", data=data)` is not correct code, because `app.post()` is actually a decorator that you should use to decorate the endpoint functions and it doesn't have `data` parameter.

As I said, you should use TestClient to imitate request to your app. In this case UploadFile object will be created automatically from the dict that you are passing as a `file` parameter. So, your endpoint function will be called with right parameters.

I think it's also possible to test the endpoint by just calling it and passing right parameters, but in this case you can't override dependencies (that is very useful feature for testing).
I don't recommend this way:
```
upload_file = UploadFile(...)
res = create_upload_file(file=upload_file)
assert res == expected_res
```

**Reply by markodjordjic** [reply]
@YuriiMotov I dont understand why are you debating with me FastAPI testing? We are deviating from a larger issue, of a potential bug in the FastAPI. I already confirmed that your example is correct, and provided the explanation that my code is clumsy because it was extracted from I larger project. For the sake of clearer explanation, here is a part of the class that does the testing:
```
import unittest
from fastapi.testclient import TestClient
from app.handling_errors import app


class TestExercise(unittest.TestCase):

    def __init__(self, methodName: str = "runTest") -> None:
        super().__init__(methodName)
 
    @classmethod
    def setUp(self) -> None:
        self.app = TestClient(app=app)
```
In summary, I am using `TestClient`.

**Reply by YuriiMotov** [reply]
> I can confirm that the proposal you suggested is working. But that is a workaround without using the UploadFile class. 

You said that code I proposed you is a **workaround**. I just wanted to show you that it's not a workaround, and explained how it works.

Can you show a code example that demonstrates FastAPI bug you are talking about?

Please check your code before posting so that there are no syntactic and logical errors in it. Because the code you posted in the initial question contains lots of errors.

**Reply by markodjordjic** [reply]
@YuriiMotov You are obviously here from wrong reasons. I have provided the all the information about potential bug, as well as why I think your proposal (not the code) is a workaround, and confirmed (already two times) that your code is correct. Its all there if you want to read.

**Reply by YuriiMotov** [reply]
Calm down, please. I'm here to help people solve their problems with FastAPI. And to learn FastAPI in depth as well.

> I have provided the all the information about potential bug

It may seem obvious to you what the bug is, but it's not clear to me.
The code you provided contains lots of errors. I can't run it and see the error message, that you mentioned in the title.

I also can see that you uses FastAPI object to perform a request instead of using TestClient (`app` variable is a global variable, that is an instance of `FastAPI` class. If you want to use TestClient object, you should change it to `self.app`).

If you want to understand, whether there is a bug in FastAPI related to UploadFile or not, please clarify your question and provide code example that we can run.

**Comment by PREPONDERANCE** [comment]
@markodjordjic I don't think the way you initialize the UploadFile is incorrect but the way you're testing it since `app` does not have a parameter called `data` and is only used for defining API endpoints.

**Reply by markodjordjic** [reply]
@PREPONDERANCE Many thanks. Can you propose a different way for the endpoint to receive the `UploadFile` class to the endpoint? I'm all ears :)

**Reply by PREPONDERANCE** [reply]
As in my experience, I'm using `Annotated` along with `File` metadata provided by FastAPI to enable file uploading. The example code will demonstrate my idea.

```python
from fastapi import File, UploadFile, APIRouter

router = APIRouter()

@router.post("/path/to/your/endpoint")
def example_code(file: Annotated[UploadFile, File]):
    pass
```

To test whether it works, one way I prefer will be accessing the openAPI docs 127.0.0.1:8000/docs (port number may differ). Find your endpoint definition and hit try. You'll see the corresponding endpoint being called in the terminal.

Another way I'd like to test this is by using tools such as postman (in VsCode, there's a plugin called ThunderClient which is very helpful as well). Turn on your local backend and call the endpoint with correct data format passed in.

**Reply by markodjordjic** [reply]
@PREPONDERANCE Thanks for proposal. So far the same error is being thrown (`Value error, Expected UploadFile, received: <class \'str\'>`) I'll keep working on it, and make a final conclusion. It seems that there is more and more evidence that there is some kind of bug. Simply put, even if the class `UploadFile` is correctly instantiated in the test, it cannot be consumed by the endpoint, regardless how the arguments are defined. I'll make further investigation, and in case you can think of something else feel free to post.

**Comment by JavierSanchezCastro** [comment]
Add a MRE: https://stackoverflow.com/help/minimal-reproducible-example

**Reply by markodjordjic** [reply]
@JavierSanchezCastro here we go:
```
import tempfile

from fastapi import FastAPI, UploadFile
from fastapi.testclient import TestClient

app = FastAPI()


@app.post("/upload_file/")
async def create_upload_file(file: UploadFile):
    
    return {"filename": file.filename}


client = TestClient(app)

def test_upload_file():
    file_name = 'test_file.txt'
    expected = {"filename": file_name}
    with tempfile.SpooledTemporaryFile(mode='wb+') as fp:
        fp.write(b'Hello World!')
        fp.seek(0)
    data = {'file': UploadFile(fp, filename=file_name)}            
    response = client.post("/upload_file/", data=data)

    actual = response.json()

    assert actual == expected

if __name__ == '__main__':

    test_upload_file()
```

**Reply by YuriiMotov** [reply]
1) You are using `fp` after exiting context manager. At that point your file is closed and `fp` can not be used. You should move next 2 lines inside context manager.
2) You are using `client.post()` in a wrong way. You can find an example of using TestClient to test endpoints that require files as parameters in the documentation of Starlette (https://www.starlette.io/testclient/). TestClient is actually a part of Starlette and FastAPI is build on top of Starlette.

Why are you so sure that you should create an instance of `UploadFile` and pass it as a parameter?
As I said before, you don't need to instantiate `UploadFile` and pass it to `client.post`. `TestClient` will do it. You should pass parameters as I showed in this answer: https://github.com/tiangolo/fastapi/discussions/11596#discussioncomment-9476998

```
    with tempfile.SpooledTemporaryFile(mode='wb+') as fp:
        fp.write(b'Hello World!')
        fp.seek(0)
        files = {'file': (file_name, fp)}
        response = client.post("/upload_file/", files=files)
        print(response.status_code)
```

Just add the following code to your endpoint function and make sure that you will get `UploadFile` object:
```
from starlette.datastructures import UploadFile as StarletteUploadFile

@app.post("/upload_file/")
async def create_upload_file(file: UploadFile):
    assert isinstance(file, StarletteUploadFile) == True
    print(f"file is an instance of {type(file)}")
    return {"filename": file.filename}
```

You will get an instance of `UploadFile` from starlette, but FastAPI's `UploadFile` is inherited from Starlette's `UploadFile` and adds some functionality to use it as a type for dependencies.

**Reply by JavierSanchezCastro** [reply]
Adding to what @YuriiMotov said before, here's an example of how I normally write test for UploadFile:

```python3
import tempfile

from fastapi import FastAPI, UploadFile
from fastapi.testclient import TestClient

app = FastAPI()


@app.post("/upload_file/")
async def create_upload_file(file: UploadFile):
    
    return {"filename": file.filename}


client = TestClient(app)

def test_upload_file():
    file_name = 'test_file.txt'
    expected = {"filename": file_name}
    with open(file_name, "rb") as f:
        files = {"file": ("test_file.txt", f, "text/plain")}
        response = client.post("/upload_file/", files=files)

    actual = response.json()
    assert actual == expected

if __name__ == '__main__':

    test_upload_file()
```

**Reply by markodjordjic** [reply]
@JavierSanchezCastro that is exactly what I needed. Thanks @JavierSanchezCastro. Just to confirm if I understood this correctly. It is not possible to "send" `UploadFile` class from the test?

**Reply by JavierSanchezCastro** [reply]
As I know, no

**Reply by markodjordjic** [reply]
@JavierSanchezCastro I'll do further testing. But if that is the case, it defeats the purpose of `UploadFile` class. What is the point of introducing a `class`, then declaring it as input data-type of a function argument, and then sending a `dict` with a `tuple`? Complete FastAPI relies on annotation of data-types, and this definitely is a deviation from that design philosophy.

**Reply by markodjordjic** [reply]
@JavierSanchezCastro But it still is worth considering is this behavior (throwing an error) as expected? Maybe it is  bug? I'll see if I can forward this to someone to clarify. @JavierSanchezCastro many thanks for the help!

**Reply by JavierSanchezCastro** [reply]
Check this one: https://github.com/tiangolo/fastapi/discussions/7470#discussioncomment-5141205

Also, you have this example using UploadFile: https://github.com/tiangolo/fastapi/discussions/7470#discussioncomment-5141218

And this comment: https://github.com/tiangolo/fastapi/discussions/7470#discussioncomment-5141220

Hope it helps!

**Reply by markodjordjic** [reply]
@JavierSanchezCastro thanks for the examples. They all point in the same direction. Declaring a function argument as 'UploadFile' data-type is not necessary (or can throw an error), since an instance of `UploadFile` class cannot be sent. The only other purpose this can serve is maybe to make a distinction from other usages of `dict`, when a `file_like` object is expected. But I guess there are better ways to annotate this. Strictly speaking function should receive only arguments of the same data-type declared in the definition of the function.

**Comment by mykhailoleskiv** [comment]
Adding the example, worked for me.
```
from typing import Annotated
from fastapi import APIRouter, UploadFile, Form
from fastapi.testclient import TestClient
from pydantic import BaseModel

router = APIRouter()
client = TestClient(app)

class Schema(BaseModel):
    file: UploadFile
    id: str

@router.post("/")
def create(data: Annotated[Schema, Form()]):
    ...

def test_create():
    filename = "test_file.csv"
    with open(filename, "rb") as f:
        response = client.post(
            "/",
            files={"file": (filename, f, "text/csv")},
            data={"id": "1"}
        }
```
